### ISSUE TITLE
Can I change the fee for the bot? Yup, BYOF

### ISSUE THREAD

**Issue by ashuralyk** [ickb/proposal#23]
the previous set fee is 0.1%, I wonder if this fee is hardcoded setup in iCKB contract

**Comment by phroi** [comment]
Hey @ashuralyk, thank you for publicly expressing your interest in iCKB as part of the integration of iCKB in NervDAO, I personally appreciate a lot!! 🙏

Please, as said before, **refrain from using market related terms**: Limit Orders are not a free market, more a conversion utility.

The current situation about fee is the following:
- Not hard-coded in the contracts (it will not change in the future, contracts are referenced by `data1`)
- Not hard-coded in the bot (for now just a basic fee check, not a percentage check)
- 0.1% hardcoded in the front-end (and for now only there)

You can change anything you want anywhere as the code is open source and in the end you are just integrating the contracts, not my suggestions on how to use them.

That said, there are three main reasons for choosing a fee not too low:
1. Disincentivizes users to abuse the Limit Order conversion utility offered in the front-end.
2. Bot conversion are capital intensive for the bot and he cannot fully stake his capital in iCKB. The bot is using his own capital to do the conversions and especially the withdrawal is capital intensive as locks up capital waiting withdrawal requests to mature.
3. Fee percentage determines for how much time the order is open in CKB to iCKB conversions. Let's make some examples. Let's say the annual NervosDAO interest rate is 2% and we use a 0.1% fee, then the bot has `2 / 365 = 0.1 / x ~ x = 0.1 * 365 / 2 = 18.25` days roughly to fulfill the order. Say we use 0.01% fee the bot has roughly 1.8 days to fulfill the order. After this time the bot will not be able to match the order as the amount requested in iCKB becomes over time bigger than the amount offered in CKB.

All this said, you can effectively chose any fee you prefer, just keep in mind that a bot will not match an order where it loses capital. Additionally, I reserve in the future the possibility of hard-coding the 0.1% in the bot and not accepting orders that do not meet it.

In short, if you are changing the fee in the front-end, please make sure that:
- All the three reasons expressed above are reasonably met.
- Your party runs a bot instance able to accept this lowered fee order, even in the case I hard-code the 0.1% in the bot.

Love & Peace, Phroi

**Comment by ashuralyk** [comment]
cool, thanks for you reply, another question, does the fee record in the data of order cell?

**Comment by phroi** [comment]
Nope, as you can see from the [data memorized in the LO cell](https://github.com/ickb/proposal#limit-order-script), the fee can only be reconstructed off-chain.

Also the fee itself varies over time:
1. At the beginning there is the fee that the user is offer, let's assume the usual 0.1%
2. At the end there is the fee that the user is paying. Let's assume the bot matches the order **after 7 days**:
 - In CKB to iCKB conversions, the final fee is 0.06% (=  0.1 - y where 2 / 365 = y / 7 ~ y = 2 * 7 / 365 = 0.04)
 - In iCKB to CKB conversions, the final fee is 0.14% (=  0.1 + y where 2 / 365 = y / 7 ~ y = 2 * 7 / 365 = 0.04)

These rough approximations are just for visualizing  the issue, so keep in mind that accountants would get bald looking at this botched math, so please don't try to incorporate it in NervDAO. Instead you can calculate the real exchange rate for any given block with the appropriate utilities: https://github.com/ickb/v1-core/blob/55755e96d10cd934f76e214903778d46b31fcc59/src/ickb_logic.ts#L399-L421

Additionally, matters are complicated by the fact that realistically a bot may have to partially match a order in multiple times to complete it. Let's say there is a LO of 1G iCKB to CKB, the bot may take many days to complete it depending on the deposits distribution in the pool and his capital availability.

Love & Peace, Phroi

**Comment by phroi** [comment]
Closing for now, feel free to reopen it if you have any other doubts 🤗

**Comment by phroi** [comment]
# Policy Change: Bring Your Own Fee

## Idea

So long that the Foundation sponsors my work and running the bot cost me zero, I can afford to use my own CKB for running a bot instance at almost zero fees for users 💪

## Bring Your Own Fee

@ashuralyk I'm changing the policy to:

> iCKB Front-ends can use any fee they want as long as the order generated get matched by a bot in a reasonable time

**TLDR**: as long the UX experience is unaffected, you can minimize the fees paid to the bot as much as you want.

Currently, as long the bot gains for each generated matching transaction [**3 times** its mining fee](https://github.com/ickb/bot/blob/6669c141c74f71c2880d7ccdab3e76a7234560a1/src/index.ts#L196), it will match all the orders contained in such transaction.

As this new policy come into effect, I'll keep monitoring the bot CKB-equivalent capital and, if I find that it loses capital, then I'll raise the bar little by little until it does not lose capital. For any such change I'll notify all the involved parties.

Losing small CKB amounts may happen in many stupid ways, but one way in particular that I have in mind is actually a future bot improvement: **the bot will actively spread clustered deposits across the 30 day cycle**. In this  way the wait time for withdrawing from a deposit at any given time is minimized. This change will happen later on when I'll complete the full migration of iCKB code-base to CCC 🔥

## Fee Structure Suggestions

I have two **alternative** suggestions:

1. Try out a few things on Testnet and see how the Bot match your orders with a new fee structure of your choice and comment on this issue, so we can discuss it.
2. [Use a `0.001% fee` while also making sure that it's at least worth `10*feeRate Sats`](https://github.com/ickb/interface/commit/ecc6f92e6b9d93b3ac1ecbdbcd358a90942ae5d8).

## `0.001% fee` Reasoning

In CKB to iCKB orders the order fee determines for how much time it is profitable to match an order. A zero fee makes the order forever un-matchable as iCKB gains value compared to CKB over time, approx 2% every year.

As [previously analyzed](https://github.com/ickb/whitepaper/issues/23#issuecomment-2791815477), the current 0.1% fee gives approximately 18 days to the bot match an CKB to iCKB order, which is an unnecessary long period.

Now, taking in account that:

- The bot is running on a stable infrastructure with minimal downtime, let's say a few hours once every month.
- The bot tries to minimize the time during which its capital is locked up in Withdrawal Requests to no more than one hour.
- Fee rate market can spike up temporarily for a few hours.

The bot may be unable to match incoming CKB to iCKB orders for a couple of hours, so we should be safe with a fee that gives more than four hours to match an order.

For example, a 0.001% fee seems perfect as it gives a bit less than 5 hours to match an order CKB to iCKB order:

`2 / (365*24) = 0.001 / x ~ x = 0.001 * 365 * 24 / 2 = 4.38 hours`

**TLDR**: the previous approx formula tell us that a NervosDAO deposit will take approx 5 hours to accrue 0.001% worth of interests.

## `10*feeRate Sats` Reasoning

Currently, as long as the bot gains for each generated matching transaction [**3 times** its mining fee](https://github.com/ickb/bot/blob/6669c141c74f71c2880d7ccdab3e76a7234560a1/src/index.ts#L196), it will match all the orders contained in such transaction.

While this minimum absolute required fee is very low, it must be accounted for. Ideally, the absolute fee should be in proportion to the median fee rate when minting the order. There could still be a successive fee rate spike preventing the order from being matched. That said, such spikes are usually temporary and later on it normalize back to a median value of fee rate

If we consider Bot matching transactions to be approx [1000 Bytes](https://explorer.nervos.org/transaction/0xcb42cb9c25ad09138b1f9d9010a827b8c3308da6d86d5eb8e29e590d5e1dedc3), then for small orders the mining fee should be around the fee rate in Sats.

So, say for example  that the current fee rate is 2000, then the mining fee incurred by the bot would be `0.00002000` CKB for matching an order. `N` times more if split in multiple transactions, as it often happens.

Just to be 100% sure that this order is actually matched and that the bot is not losing funds in stupid ways (like a constant stream of minuscule orders), we can multiply this minimal mining fee by `10`, hence the proposed  `10*feeRate Sats`.

## Conclusion

I'll keep this issue open as a form of documentation and fee discussion place, feel free to ask any questions 🤗

Love & Peace, Phroi %79
